pkg/daemon/route/utils.go

```diff
@@ -402,19 +402,52 @@ func ensureRoutesForVlanSubnet(forwardLink netlink.Link, cidr *net.IPNet, gatewa
 	return nil
 }
 
-func ensureRoutesForBGPSubnet(forwardLink netlink.Link, cidr *net.IPNet, table int, gateway net.IP) error {
-	// don't use onlink flag in case the gateway is not a reachable next hop
-	defaultRoute := &netlink.Route{
-		LinkIndex: forwardLink.Attrs().Index,
-		Table:     table,
-		Scope:     netlink.SCOPE_UNIVERSE,
-		Gw:        gateway,
+func ensureRoutesForBGPSubnet(forwardLink netlink.Link, cidr *net.IPNet, gateway net.IP, table, family int) error {
+	// default route is always needed
+	var defaultRoute *netlink.Route
+	var err error
+
+	if gateway == nil {
+		// copy the origin node default route in bgp subnet table
+		defaultRoute, err = daemonutils.GetDefaultRoute(family)
+		if err != nil {
+			return fmt.Errorf("failed to get default route in mian table: %v", err)
+		}
+		defaultRoute.Table = table
+	} else {
+		// TODO: support multiple bgp gateway
+		// don't use onlink flag in case the gateway is not a reachable next hop
+		defaultRoute = &netlink.Route{
+			LinkIndex: forwardLink.Attrs().Index,
+			Table:     table,
+			Scope:     netlink.SCOPE_UNIVERSE,
+			Gw:        gateway,
+		}
 	}
 
 	if err := netlink.RouteReplace(defaultRoute); err != nil {
 		return fmt.Errorf("failed to add bgp subnet %v default route %v: %v", cidr.String(), defaultRoute.String(), err)
 	}
 
+	// Because `ip route replace` will not delete default route if gateway changed, we need to delete it additionally.
+	routeList, err := netlink.RouteListFiltered(family, &netlink.Route{
+		Table: table,
+	}, netlink.RT_FILTER_TABLE)
+	if err != nil {
+		return fmt.Errorf("failed to list route for table %v: %v", table, err)
+	}
+
+	for _, route := range routeList {
+		// cannot use route.Equal() because of empty fields
+		if daemonutils.IsDefaultRoute(&route, family) &&
+			// TODO: support multiple bgp gateway
+			(!route.Gw.Equal(defaultRoute.Gw) || route.LinkIndex != defaultRoute.LinkIndex) {
```

Comment:
can we restrict the deletion scope here, in other words, is the judgement for the route deletion safe enough?

Reply:
Yes, it is safe, only daefault routes with different gateway and interface will be deleted.